File: Problems/No.82_Remove_Duplicates_from_Sorted_List_II/remove_duplicates_from_sorted_list_II.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* deleteDuplicates(ListNode* head) {
        
        // parameters
        ListNode *newHead = new ListNode(0, head);
        ListNode *dFirst = newHead, *dLast;
        bool NULL_check;

        // start computing
        if ((head == NULL) || (head->next == NULL)) return head;
        else dLast = dFirst->next->next;
        while (dLast != NULL) {

            NULL_check = false;
            if (dFirst->next->val == dLast->val) {

                while (dFirst->next->val == dLast->val) {

                    dLast = dLast->next;
                    if (dLast == NULL) {

                        NULL_check = true;
                        break;

                    }

                }
                dFirst->next = dLast;

            }
            else dFirst = dFirst->next;

            if (!NULL_check) dLast = dLast->next;

        }

        return newHead->next;

    }
};

File: Problems/No.82_Remove_Duplicates_from_Sorted_List_II/remove_duplicates_from_sorted_list_II.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution(object):
    def deleteDuplicates(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """

        # parameters
        newHead = ListNode(0, head)
        dFirst, dLast = newHead, None
        None_check = None

        # start computing
        if ((head == None) or (head.next == None)):
            return head
        else:
            dLast = dFirst.next.next
        while (dLast):
            None_check = False
            if (dFirst.next.val == dLast.val):
                while (dFirst.next.val == dLast.val):
                    dLast = dLast.next
                    if not dLast:
                        None_check = True
                        break
                dFirst.next = dLast
            else:
                dFirst = dFirst.next
            if not None_check:
                dLast = dLast.next

        return newHead.next
